yt/frontends/parthenon/io.py

```diff
@@ -0,0 +1,107 @@
+from itertools import groupby
+
+import numpy as np
+
+from yt.utilities.io_handler import BaseIOHandler
+from yt.utilities.logger import ytLogger as mylog
+
+
+# http://stackoverflow.com/questions/2361945/detecting-consecutive-integers-in-a-list
+def grid_sequences(grids):
+    g_iter = sorted(grids, key=lambda g: g.id)
+    for _, g in groupby(enumerate(g_iter), lambda i_x1: i_x1[0] - i_x1[1].id):
+        seq = [v[1] for v in g]
+        yield seq
+
+
+ii = [0, 1, 0, 1, 0, 1, 0, 1]
+jj = [0, 0, 1, 1, 0, 0, 1, 1]
+kk = [0, 0, 0, 0, 1, 1, 1, 1]
+
+
+class IOHandlerParthenon(BaseIOHandler):
+    _particle_reader = False
+    _dataset_type = "parthenon"
+
+    def __init__(self, ds):
+        super().__init__(ds)
+        self._handle = ds._handle
+
+    def _read_particles(
+        self, fields_to_read, type, args, grid_list, count_list, conv_factors
+    ):
+        pass
+
+    def _read_fluid_selection(self, chunks, selector, fields, size):
+        chunks = list(chunks)
+        if any((ftype != "parthenon" for ftype, fname in fields)):
+            raise NotImplementedError
+        f = self._handle
+        rv = {}
+        for field in fields:
+            # Always use *native* 64-bit float.
+            rv[field] = np.empty(size, dtype="=f8")
+        ng = sum(len(c.objs) for c in chunks)
+        mylog.debug(
+            "Reading %s cells of %s fields in %s blocks",
+            size,
+            [f2 for f1, f2 in fields],
+            ng,
+        )
+        last_dname = None
+        for field in fields:
```

Comment:
So most (all other?) frontends loop over chunks then fields, e.g., 

``` python
for chunk in chunks:
     # do stuff
     for field in fields:
            # get this fields data for this chunk
            # apply selector
```
The main reason being that yt's parallel chunking works by sending chunks to different processors where field data is selected. So if you eventually want the Parthenon frontend to work with yt's parallel operations, you'll want to invert the loop order here. I don't think this needs to be done for this PR, but a refactor in the future might be warranted...

Reply:
I have tried to use yt's parallel functionality before, and have not been able to get it to work, so this might explain why. It would be ideal if this could work. Is inverting the loop order the only change that's needed to get it working?

Reply:
So I think I should have said "to work more efficiently with yt's parallel operations" -- generally it's better to process all the fields within a chunk rather than process every chunk for each field. But after thinking more, it might actually not be too important here since it's reading from a single hdf file -- when chunks are spread across files if you put the chunk loop on the inside you'd be incurring extra cost from opening/closing the same file multiple times. 

In terms of getting parallel operations to work -- I don't **think** there should be more changes. When you run a parallel-enabled function with MPI, I believe the `chunks` argument should come in to this function with only the chunks associated with the MPI process running the function. I can test that out though, I've been meaning to get back into some parallel indexing related work....

Reply:
In that case, I'm currently leaning towards leaving as it is and keep this in mind in case we run into issues (which we kind of are, but processing 3TB dumps is a story for another day).